crates/matrix-sdk/tests/integration/room_preview.rs

```diff
@@ -0,0 +1,139 @@
+use assert_matches2::assert_matches;
+use matrix_sdk::{config::SyncSettings, test_utils::logged_in_client_with_server, Error};
+use matrix_sdk_base::RoomState;
+use matrix_sdk_test::{
+    async_test, InvitedRoomBuilder, JoinedRoomBuilder, KnockedRoomBuilder, SyncResponseBuilder,
+};
+use ruma::{room_id, space::SpaceRoomJoinRule, RoomId};
+use serde_json::json;
+use wiremock::{
+    matchers::{header, method, path_regex},
+    Mock, MockServer, ResponseTemplate,
+};
+
+use crate::mock_sync;
+
+#[async_test]
+async fn test_room_preview_leave_invited() {
+    let (client, server) = logged_in_client_with_server().await;
+    let room_id = room_id!("!room:localhost");
+
+    let mut sync_builder = SyncResponseBuilder::new();
+    sync_builder.add_invited_room(InvitedRoomBuilder::new(room_id));
+
+    mock_sync(&server, sync_builder.build_json_sync_response(), None).await;
+    client.sync_once(SyncSettings::default()).await.unwrap();
+    server.reset().await;
+
+    mock_leave(room_id, &server).await;
+
+    let room_preview = client.get_room_preview(room_id.into(), Vec::new()).await.unwrap();
+    assert_eq!(room_preview.state.unwrap(), RoomState::Invited);
+
+    room_preview.leave().await.unwrap();
+
+    assert_eq!(client.get_room(room_id).unwrap().state(), RoomState::Left);
+}
+
+#[async_test]
+async fn test_room_preview_leave_knocked() {
```

Comment:
With the suggestions I've made above, I think that most of these tests could go away. Unless we didn't have tests to leave a knocked room, maybe?